sdk/modelsrepository/azure-iot-modelsrepository/src/main/java/com/azure/iot/modelsrepository/implementation/FileModelFetcher.java

```diff
@@ -0,0 +1,85 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.iot.modelsrepository.implementation;
+
+import com.azure.core.exception.AzureException;
+import com.azure.core.exception.ServiceResponseException;
+import com.azure.core.util.Context;
+import com.azure.core.util.logging.ClientLogger;
+import com.azure.iot.modelsrepository.DtmiConventions;
+import com.azure.iot.modelsrepository.ModelsDependencyResolution;
+import com.azure.iot.modelsrepository.implementation.models.FetchResult;
+import reactor.core.publisher.Mono;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.LinkedList;
+import java.util.Queue;
+
+/**
+ * The {@link FileModelFetcher} is an implementation of {@link ModelFetcher} interface
+ * for supporting local filesystem based model content fetching.
+ */
+class FileModelFetcher implements ModelFetcher {
+
+    private final ClientLogger logger;
+
+    /**
+     * Creates an instance of {@link FileModelFetcher}
+     */
+    public FileModelFetcher(ClientLogger logger) {
+        this.logger = logger;
+    }
+
+    @Override
+    public Mono<FetchResult> fetchAsync(String dtmi, URI repositoryUri, ModelsDependencyResolution resolutionOption, Context context) {
+        Queue<String> work = new LinkedList<>();
+
+        try {
+            if (resolutionOption == ModelsDependencyResolution.TRY_FROM_EXPANDED) {
+                work.add(getPath(dtmi, repositoryUri, true));
+            }
+
+            work.add(getPath(dtmi, repositoryUri, false));
+        } catch (Exception e) {
+            return Mono.error(new AzureException(e));
+        }
+
+        String fnfError = "";
+        while (work.size() != 0) {
+            String tryContentPath = work.poll();
+
+            Path path = Path.of(new File(tryContentPath).getPath());
+
+            logger.info(String.format(StandardStrings.FetchingModelContent, path.toString()));
+
+            if (Files.exists(path)) {
+                try {
+                    return Mono.just(
+                        new FetchResult()
+                            .setDefinition(new String(Files.readAllBytes(path)))
+                            .setPath(tryContentPath));
+                } catch (IOException e) {
+                    return Mono.error(new AzureException(e));
+                }
+            }
+
+            logger.error(String.format(StandardStrings.ErrorFetchingModelContent, path.toString()));
+
+            fnfError = String.format(StandardStrings.ErrorFetchingModelContent, tryContentPath);
+        }
```

Comment:
This is not async. The method returns async publisher so this should be done asynchronously.

Reply:
This just reads stuff off of the file system, can you elaborate how that can be done async?

Reply:
fixed.